testplan/testing/base.py

```diff
@@ -366,10 +389,86 @@ def _record_teardown_start(self):
     def _record_teardown_end(self):
         self.report.timer.end("teardown")
 
+    def _init_context(self):
+        if callable(self.cfg.initial_context):
+            self.resources._initial_context = self.cfg.initial_context()
+        else:
+            self.resources._initial_context = self.cfg.initial_context
+
+    def _build_environment(self, delay: bool):
+        """
+        This method will execute at most twice, once with delay=True during __init__,
+        and then with delay=False during _run. It will be no-op if called more
+        than once in interactive mode.
+        :param delay: delay eval if self.cfg.environment is a callable
+        """
+
+        if self._env_built:
+            return
+
+        if delay:
+            if callable(self.cfg.environment):
+                return
+            else:
+                drivers = self.cfg.environment
+        else:
+            if callable(self.cfg.environment):
+                drivers = self.cfg.environment()
+            else:
+                return
+
+        for driver in drivers:
+            driver.parent = self
+            driver.cfg.parent = self.cfg
+            self.resources.add(driver)
+        self._env_built = True
+
+    def _set_dependencies(self):
+        if callable(self.cfg.dependencies):
+            deps = parse_dependency(self.cfg.dependencies())
+        else:
+            deps = self.cfg.dependencies
+        if deps:
+            self.resources.set_dependency(deps)
+
     def pre_resource_steps(self):
         """Runnable steps to be executed before environment starts."""
         self._add_step(self._record_setup_start)
 
+        self._add_step(self._init_context)
+        self._add_step(self._build_environment, delay=False)
+        self._add_step(self._set_dependencies)
+
+    def start_resource_steps(self):
```

Comment:
I know this naming conforms to the rest, but it always tricks me thinking that it actually execute those steps. This is a nice to have, but what if we rename these to something enqueue_..._steps ?

Reply:
agree